folsom/src/main/java/com/spotify/folsom/client/Utils.java

```diff
@@ -72,4 +74,43 @@ public static Throwable unwrap(final Throwable e) {
     }
     return future.whenCompleteAsync((t, throwable) -> {}, executor);
   }
+
+  /**
+   * Checks if an executor is direct or not. Direct means that the executor executes work on the
+   * same thread as the work was submitted on.
+   *
+   * @param executor may not be null
+   * @return true if the executor is a direct executor, otherwise false
+   */
+  static boolean isDirectExecutor(Executor executor) {
+    requireNonNull(executor);
+    RunnableWithThread runnableWithThread = new RunnableWithThread();
+    try {
+      executor.execute(runnableWithThread);
+
+      // We have the following cases
+      // 1) It is a direct executor, so runnableWithThread.thread will be set to the current thread.
+      //    We will correctly return true
+      // 2) It is not a direct executor and runnableWithThread.thread is still null when we check
+      // it.
+      //    We correctly return false
+      // 3) It is not a direct executor but the runnableWithThread.thread somehow managed to get set
+      //    before we check it. In any case, it can't be referencing the same thread we are in
+      //    We correctly return false
+
+      return runnableWithThread.thread == Thread.currentThread();
+    } catch (Exception e) {
+      // If the executor throws any exception, it can not be a direct executor
```

Comment:
Is this as safe assumption? Looking at the Guava DirectExecutorService, it can throw exceptions here, but only if the executor is shutting down, which is probably fine either way. But, would we want to assume other direct executors would behave the same way?

Reply:
No, probably not a safe assumption. It depends a bit on what the function is used for. If it's only for optimization purposes and sanity checks it may still be good enough